frame/staking/src/pallet/mod.rs

```diff
@@ -612,6 +630,11 @@ pub mod pallet {
 		fn on_runtime_upgrade() -> Weight {
 			if StorageVersion::<T>::get() == Releases::V6_0_0 {
 				migrations::v7::migrate::<T>()
+			} else if StorageVersion::<T>::get() == Releases::V7_0_0 {
+				// NOTE: this migration assumes that pallet-bags-list is being used as the
+				// [`SortedListProvider`]. If not using pallet-bags-list the runtime author should
+				// carefully consider if this migration is necessary.
+				migrations::v8::migrate::<T>()
```

Comment:
there is no way to opt-out of the hook. If runtime author needs to be careful about this migration then it shouldn't be a hidden message in the source code of the pallet. I think the best is to not put this in the hook but instead have clear message in the PR about the breaking change, and allow user to trigger this migration manually when they define the runtime upgrade of the runtime.

That said for me this migration doesn't require to use pallet-bags-list, in case someone use another implementation it will work fine, UseNominatorMap will no-op for example.
The only warning is that it could potentially blow up with too many writes maybe.

Reply:
not sure why this was closed, but it is important, and we should not merge until the migration is out of the pallet.